src/Forms/SingleSelectField.php

```diff
@@ -50,6 +47,17 @@ public function getSchemaDataDefaults()
         return $data;
     }
 
+    public function getValueForValidation(): mixed
+    {
+        $value = $this->getValue();
+        if ($value === '' && $this->getHasEmptyDefault()) {
+            return null;
+        } elseif (is_string($value) && is_numeric($value) && $this->getAllSourceValuesAreIntType()) {
+            return (int) $value;
+        }
```

Comment:
I think this is an incomplete solution. If I have

```php
DropdownField::create('MyField', null, ['one', 'two' => 'two', 'three']);
// or
DropdownField::create('MyField', null, [1 => 'one', 'two' => 'two',  '3' => 'three']);
```

The integer keys are not being converted to `int` so I get a "Not an allowed value" error if selecting "one" from the dropdown. Or at least I assume that's what's causing the error.

This method probably needs to check against the keys in `$this->getSource()` and only cast to int if it's an int in the source, or something along those lines.

Probably also needs an inline comment explaining why the cast is happening - without the context of this PR it'll be pretty hard to just look at it and understand why it's doing what it's doing.

Reply:
Have redone solution, remove SelectField::getAllSourceValuesAreIntType() and created SelectField::castSubmittedValue() instead